###  Importing All the basic necessary Libraries 

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from sklearn.model_selection import GridSearchCV



### 1️⃣ Stage 1 : Data Understanding & Preprocessing 

In [3]:
df = pd.read_csv(r"C:\Users\MOHAMMED YAZIN N\Downloads\CarPrice_Assignment.csv")
df

car_ID  symboling                   CarName fueltype aspiration  \
0         1          3        alfa-romero giulia      gas        std   
1         2          3       alfa-romero stelvio      gas        std   
2         3          1  alfa-romero Quadrifoglio      gas        std   
3         4          2               audi 100 ls      gas        std   
4         5          2                audi 100ls      gas        std   
..      ...        ...                       ...      ...        ...   
200     201         -1           volvo 145e (sw)      gas        std   
201     202         -1               volvo 144ea      gas      turbo   
202     203         -1               volvo 244dl      gas        std   
203     204         -1                 volvo 246   diesel      turbo   
204     205         -1               volvo 264gl      gas      turbo   

    doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0          two  convertible        rwd          front       88.6  ...   
1          two  convertible        rwd          front       88.6  ...   
2          two    hatchback        rwd          front       94.5  ...   
3         four        sedan        fwd          front       99.8  ...   
4         four        sedan        4wd          front       99.4  ...   
..         ...          ...        ...            ...        ...  ...   
200       four        sedan        rwd          front      109.1  ...   
201       four        sedan        rwd          front      109.1  ...   
202       four        sedan        rwd          front      109.1  ...   
203       four        sedan        rwd          front      109.1  ...   
204       four        sedan        rwd          front      109.1  ...   

     enginesize  fuelsystem  boreratio  stroke compressionratio horsepower  \
0           130        mpfi       3.47    2.68              9.0        111   
1           130        mpfi       3.47    2.68              9.0        111   
2           152        mpfi       2.68    3.47              9.0        154   
3           109        mpfi       3.19    3.40             10.0        102   
4           136        mpfi       3.19    3.40              8.0        115   
..          ...         ...        ...     ...              ...        ...   
200         141        mpfi       3.78    3.15              9.5        114   
201         141        mpfi       3.78    3.15              8.7        160   
202         173        mpfi       3.58    2.87              8.8        134   
203         145         idi       3.01    3.40             23.0        106   
204         141        mpfi       3.78    3.15              9.5        114   

     peakrpm citympg  highwaympg    price  
0       5000      21          27  13495.0  
1       5000      21          27  16500.0  
2       5000      19          26  16500.0  
3       5500      24          30  13950.0  
4       5500      18          22  17450.0  
..       ...     ...         ...      ...  
200     5400      23          28  16845.0  
201     5300      19          25  19045.0  
202     5500      18          23  21485.0  
203     4800      26          27  22470.0  
204     5400      19          25  22625.0  

[205 rows x 26 columns]

### Exploratory Data Analysis

In [4]:
df.shape

(205, 26)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

In [6]:
## Statistical Summary 
df.describe()

car_ID   symboling   wheelbase   carlength    carwidth   carheight  \
count  205.000000  205.000000  205.000000  205.000000  205.000000  205.000000   
mean   103.000000    0.834146   98.756585  174.049268   65.907805   53.724878   
std     59.322565    1.245307    6.021776   12.337289    2.145204    2.443522   
min      1.000000   -2.000000   86.600000  141.100000   60.300000   47.800000   
25%     52.000000    0.000000   94.500000  166.300000   64.100000   52.000000   
50%    103.000000    1.000000   97.000000  173.200000   65.500000   54.100000   
75%    154.000000    2.000000  102.400000  183.100000   66.900000   55.500000   
max    205.000000    3.000000  120.900000  208.100000   72.300000   59.800000   

        curbweight  enginesize   boreratio      stroke  compressionratio  \
count   205.000000  205.000000  205.000000  205.000000        205.000000   
mean   2555.565854  126.907317    3.329756    3.255415         10.142537   
std     520.680204   41.642693    0.270844    0.313597          3.972040   
min    1488.000000   61.000000    2.540000    2.070000          7.000000   
25%    2145.000000   97.000000    3.150000    3.110000          8.600000   
50%    2414.000000  120.000000    3.310000    3.290000          9.000000   
75%    2935.000000  141.000000    3.580000    3.410000          9.400000   
max    4066.000000  326.000000    3.940000    4.170000         23.000000   

       horsepower      peakrpm     citympg  highwaympg         price  
count  205.000000   205.000000  205.000000  205.000000    205.000000  
mean   104.117073  5125.121951   25.219512   30.751220  13276.710571  
std     39.544167   476.985643    6.542142    6.886443   7988.852332  
min     48.000000  4150.000000   13.000000   16.000000   5118.000000  
25%     70.000000  4800.000000   19.000000   25.000000   7788.000000  
50%     95.000000  5200.000000   24.000000   30.000000  10295.000000  
75%    116.000000  5500.000000   30.000000   34.000000  16503.000000  
max    288.000000  6600.000000   49.000000   54.000000  45400.000000

In [7]:
# Check missing values
print(df.isnull().sum())

car_ID              0
symboling           0
CarName             0
fueltype            0
aspiration          0
doornumber          0
carbody             0
drivewheel          0
enginelocation      0
wheelbase           0
carlength           0
carwidth            0
carheight           0
curbweight          0
enginetype          0
cylindernumber      0
enginesize          0
fuelsystem          0
boreratio           0
stroke              0
compressionratio    0
horsepower          0
peakrpm             0
citympg             0
highwaympg          0
price               0
dtype: int64


In [8]:
## Unique Values in Each columns 
df.nunique()

car_ID              205
symboling             6
CarName             147
fueltype              2
aspiration            2
doornumber            2
carbody               5
drivewheel            3
enginelocation        2
wheelbase            53
carlength            75
carwidth             44
carheight            49
curbweight          171
enginetype            7
cylindernumber        7
enginesize           44
fuelsystem            8
boreratio            38
stroke               37
compressionratio     32
horsepower           59
peakrpm              23
citympg              29
highwaympg           30
price               189
dtype: int64

In [9]:
## visulisiing the data types 
df.dtypes

car_ID                int64
symboling             int64
CarName              object
fueltype             object
aspiration           object
doornumber           object
carbody              object
drivewheel           object
enginelocation       object
wheelbase           float64
carlength           float64
carwidth            float64
carheight           float64
curbweight            int64
enginetype           object
cylindernumber       object
enginesize            int64
fuelsystem           object
boreratio           float64
stroke              float64
compressionratio    float64
horsepower            int64
peakrpm               int64
citympg               int64
highwaympg            int64
price               float64
dtype: object

In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df.head()

car_ID  symboling                   CarName fueltype aspiration doornumber  \
0       1          3        alfa-romero giulia      gas        std        two   
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  enginesize  \
0  convertible        rwd          front       88.6  ...         130   
1  convertible        rwd          front       88.6  ...         130   
2    hatchback        rwd          front       94.5  ...         152   
3        sedan        fwd          front       99.8  ...         109   
4        sedan        4wd          front       99.4  ...         136   

   fuelsystem  boreratio  stroke compressionratio horsepower  peakrpm citympg  \
0        mpfi       3.47    2.68              9.0        111     5000      21   
1        mpfi       3.47    2.68              9.0        111     5000      21   
2        mpfi       2.68    3.47              9.0        154     5000      19   
3        mpfi       3.19    3.40             10.0        102     5500      24   
4        mpfi       3.19    3.40              8.0        115     5500      18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

In [12]:
df.tail()

car_ID  symboling          CarName fueltype aspiration doornumber  \
200     201         -1  volvo 145e (sw)      gas        std       four   
201     202         -1      volvo 144ea      gas      turbo       four   
202     203         -1      volvo 244dl      gas        std       four   
203     204         -1        volvo 246   diesel      turbo       four   
204     205         -1      volvo 264gl      gas      turbo       four   

    carbody drivewheel enginelocation  wheelbase  ...  enginesize  fuelsystem  \
200   sedan        rwd          front      109.1  ...         141        mpfi   
201   sedan        rwd          front      109.1  ...         141        mpfi   
202   sedan        rwd          front      109.1  ...         173        mpfi   
203   sedan        rwd          front      109.1  ...         145         idi   
204   sedan        rwd          front      109.1  ...         141        mpfi   

     boreratio  stroke compressionratio horsepower  peakrpm citympg  \
200       3.78    3.15              9.5        114     5400      23   
201       3.78    3.15              8.7        160     5300      19   
202       3.58    2.87              8.8        134     5500      18   
203       3.01    3.40             23.0        106     4800      26   
204       3.78    3.15              9.5        114     5400      19   

     highwaympg    price  
200          28  16845.0  
201          25  19045.0  
202          23  21485.0  
203          27  22470.0  
204          25  22625.0  

[5 rows x 26 columns]

In [13]:
df.columns

Index(['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration',
       'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase',
       'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype',
       'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

In [14]:
# Extract brand from 'CarName'
df['CarBrand'] = df['CarName'].apply(lambda x: x.split()[0].lower())
df.drop(['CarName'], axis=1, inplace=True)

# Fix inconsistent brand names
df['CarBrand'] = df['CarBrand'].replace({'vw': 'volkswagen', 'vokswagen': 'volkswagen', 'maxda': 'mazda', 'porcshce': 'porsche', 'toyouta': 'toyota'})
df['CarBrand'].unique()

# Drop car_ID column (not useful for modeling)
df.drop(['car_ID'], axis=1, inplace=True)

In [15]:
df

symboling fueltype aspiration doornumber      carbody drivewheel  \
0            3      gas        std        two  convertible        rwd   
1            3      gas        std        two  convertible        rwd   
2            1      gas        std        two    hatchback        rwd   
3            2      gas        std       four        sedan        fwd   
4            2      gas        std       four        sedan        4wd   
..         ...      ...        ...        ...          ...        ...   
200         -1      gas        std       four        sedan        rwd   
201         -1      gas      turbo       four        sedan        rwd   
202         -1      gas        std       four        sedan        rwd   
203         -1   diesel      turbo       four        sedan        rwd   
204         -1      gas      turbo       four        sedan        rwd   

    enginelocation  wheelbase  carlength  carwidth  ...  fuelsystem  \
0            front       88.6      168.8      64.1  ...        mpfi   
1            front       88.6      168.8      64.1  ...        mpfi   
2            front       94.5      171.2      65.5  ...        mpfi   
3            front       99.8      176.6      66.2  ...        mpfi   
4            front       99.4      176.6      66.4  ...        mpfi   
..             ...        ...        ...       ...  ...         ...   
200          front      109.1      188.8      68.9  ...        mpfi   
201          front      109.1      188.8      68.8  ...        mpfi   
202          front      109.1      188.8      68.9  ...        mpfi   
203          front      109.1      188.8      68.9  ...         idi   
204          front      109.1      188.8      68.9  ...        mpfi   

     boreratio stroke compressionratio  horsepower peakrpm  citympg  \
0         3.47   2.68              9.0         111    5000       21   
1         3.47   2.68              9.0         111    5000       21   
2         2.68   3.47              9.0         154    5000       19   
3         3.19   3.40             10.0         102    5500       24   
4         3.19   3.40              8.0         115    5500       18   
..         ...    ...              ...         ...     ...      ...   
200       3.78   3.15              9.5         114    5400       23   
201       3.78   3.15              8.7         160    5300       19   
202       3.58   2.87              8.8         134    5500       18   
203       3.01   3.40             23.0         106    4800       26   
204       3.78   3.15              9.5         114    5400       19   

     highwaympg    price     CarBrand  
0            27  13495.0  alfa-romero  
1            27  16500.0  alfa-romero  
2            26  16500.0  alfa-romero  
3            30  13950.0         audi  
4            22  17450.0         audi  
..          ...      ...          ...  
200          28  16845.0        volvo  
201          25  19045.0        volvo  
202          23  21485.0        volvo  
203          27  22470.0        volvo  
204          25  22625.0        volvo  

[205 rows x 25 columns]

### Seperating Categorical and numerical features of the dataset 

In [16]:
# Identify categorical features
cat_cols = df.select_dtypes(include='object').columns
cat_cols



Index(['fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel',
       'enginelocation', 'enginetype', 'cylindernumber', 'fuelsystem',
       'CarBrand'],
      dtype='object')

### Encoding Categorical Variables


In [17]:
# Performing One-Hot Encoding 
df_encoded = pd.get_dummies(df, columns=cat_cols,drop_first=True,dtype=int)
df_encoded

symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
0            3       88.6      168.8      64.1       48.8        2548   
1            3       88.6      168.8      64.1       48.8        2548   
2            1       94.5      171.2      65.5       52.4        2823   
3            2       99.8      176.6      66.2       54.3        2337   
4            2       99.4      176.6      66.4       54.3        2824   
..         ...        ...        ...       ...        ...         ...   
200         -1      109.1      188.8      68.9       55.5        2952   
201         -1      109.1      188.8      68.8       55.5        3049   
202         -1      109.1      188.8      68.9       55.5        3012   
203         -1      109.1      188.8      68.9       55.5        3217   
204         -1      109.1      188.8      68.9       55.5        3062   

     enginesize  boreratio  stroke  compressionratio  ...  CarBrand_nissan  \
0           130       3.47    2.68               9.0  ...                0   
1           130       3.47    2.68               9.0  ...                0   
2           152       2.68    3.47               9.0  ...                0   
3           109       3.19    3.40              10.0  ...                0   
4           136       3.19    3.40               8.0  ...                0   
..          ...        ...     ...               ...  ...              ...   
200         141       3.78    3.15               9.5  ...                0   
201         141       3.78    3.15               8.7  ...                0   
202         173       3.58    2.87               8.8  ...                0   
203         145       3.01    3.40              23.0  ...                0   
204         141       3.78    3.15               9.5  ...                0   

     CarBrand_peugeot  CarBrand_plymouth  CarBrand_porsche  CarBrand_renault  \
0                   0                  0                 0                 0   
1                   0                  0                 0                 0   
2                   0                  0                 0                 0   
3                   0                  0                 0                 0   
4                   0                  0                 0                 0   
..                ...                ...               ...               ...   
200                 0                  0                 0                 0   
201                 0                  0                 0                 0   
202                 0                  0                 0                 0   
203                 0                  0                 0                 0   
204                 0                  0                 0                 0   

     CarBrand_saab  CarBrand_subaru  CarBrand_toyota  CarBrand_volkswagen  \
0                0                0                0                    0   
1                0                0                0                    0   
2                0                0                0                    0   
3                0                0                0                    0   
4                0                0                0                    0   
..             ...              ...              ...                  ...   
200              0                0                0                    0   
201              0                0                0                    0   
202              0                0                0                    0   
203              0                0                0                    0   
204              0                0                0                    0   

     CarBrand_volvo  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  
..              ...  
200               1  
201               1  
202               1  
203               1  
204               1  

[205 rows x 65 columns]

### Defining features and target variable

In [18]:
# Features and target
X = df_encoded.drop('price', axis=1)
y = df_encoded['price']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [19]:
X_train

symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
66           0      104.9      175.0      66.1       54.4        2700   
111          0      107.9      186.7      68.4       56.7        3075   
153          0       95.7      169.7      63.6       59.1        2280   
96           1       94.5      165.3      63.8       54.5        1971   
38           0       96.5      167.5      65.2       53.3        2289   
..         ...        ...        ...       ...        ...         ...   
106          1       99.2      178.5      67.9       49.7        3139   
14           1      103.5      189.0      66.9       55.7        3055   
92           1       94.5      165.3      63.8       54.5        1938   
179          3      102.9      183.5      67.7       52.0        3016   
102          0      100.4      184.6      66.5       56.1        3296   

     enginesize  boreratio  stroke  compressionratio  ...  CarBrand_nissan  \
66          134       3.43    3.64              22.0  ...                0   
111         120       3.46    2.19               8.4  ...                0   
153          92       3.05    3.03               9.0  ...                0   
96           97       3.15    3.29               9.4  ...                1   
38          110       3.15    3.58               9.0  ...                0   
..          ...        ...     ...               ...  ...              ...   
106         181       3.43    3.27               9.0  ...                1   
14          164       3.31    3.19               9.0  ...                0   
92           97       3.15    3.29               9.4  ...                1   
179         171       3.27    3.35               9.3  ...                0   
102         181       3.43    3.27               9.0  ...                1   

     CarBrand_peugeot  CarBrand_plymouth  CarBrand_porsche  CarBrand_renault  \
66                  0                  0                 0                 0   
111                 1                  0                 0                 0   
153                 0                  0                 0                 0   
96                  0                  0                 0                 0   
38                  0                  0                 0                 0   
..                ...                ...               ...               ...   
106                 0                  0                 0                 0   
14                  0                  0                 0                 0   
92                  0                  0                 0                 0   
179                 0                  0                 0                 0   
102                 0                  0                 0                 0   

     CarBrand_saab  CarBrand_subaru  CarBrand_toyota  CarBrand_volkswagen  \
66               0                0                0                    0   
111              0                0                0                    0   
153              0                0                1                    0   
96               0                0                0                    0   
38               0                0                0                    0   
..             ...              ...              ...                  ...   
106              0                0                0                    0   
14               0                0                0                    0   
92               0                0                0                    0   
179              0                0                1                    0   
102              0                0                0                    0   

     CarBrand_volvo  
66                0  
111               0  
153               0  
96                0  
38                0  
..              ...  
106               0  
14                0  
92                0  
179               0  
102               0  

[164 rows x 64 columns]

In [20]:
X_test

symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
15           0      103.5      189.0      66.9       55.7        3230   
9            0       99.5      178.2      67.9       52.0        3053   
100          0       97.2      173.4      65.2       54.7        2302   
132          3       99.1      186.6      66.5       56.1        2658   
68          -1      110.0      190.9      70.3       58.7        3750   
95           1       94.5      165.6      63.8       53.3        2028   
159          0       95.7      166.3      64.4       52.8        2275   
162          0       95.7      166.3      64.4       52.8        2140   
147          0       97.0      173.5      65.4       53.0        2455   
182          2       97.3      171.7      65.5       55.7        2261   
191          0      100.4      180.2      66.9       55.1        2661   
164          1       94.5      168.7      64.0       52.6        2204   
65           0      104.9      175.0      66.1       54.4        2670   
175         -1      102.4      175.6      66.5       53.9        2414   
73           0      120.9      208.1      71.7       56.7        3900   
152          1       95.7      158.7      63.6       54.5        2015   
18           2       88.4      141.1      60.3       53.2        1488   
82           3       95.9      173.2      66.3       50.2        2833   
86           1       96.3      172.4      65.4       51.6        2405   
143          0       97.2      172.0      65.4       52.5        2340   
60           0       98.8      177.8      66.5       55.5        2410   
101          0      100.4      181.7      66.5       55.1        3095   
98           2       95.1      162.4      63.8       53.3        2008   
30           2       86.6      144.6      63.9       50.8        1713   
25           1       93.7      157.3      63.8       50.6        1989   
16           0      103.5      193.8      67.9       53.7        3380   
168          2       98.4      176.2      65.6       52.0        2536   
195         -1      104.3      188.8      67.2       57.5        3034   
97           1       94.5      170.2      63.8       53.5        2037   
194         -2      104.3      188.8      67.2       56.2        2912   
67          -1      110.0      190.9      70.3       56.5        3515   
120          1       93.7      157.3      63.8       50.6        1967   
154          0       95.7      169.7      63.6       59.1        2290   
202         -1      109.1      188.8      68.9       55.5        3012   
79           1       93.0      157.3      63.8       50.8        2145   
69           0      106.7      187.5      70.3       54.9        3495   
145          0       97.0      172.0      65.4       54.3        2510   
55           3       95.3      169.0      65.7       49.6        2380   
45           0       94.5      155.9      63.6       52.0        1909   
84           3       95.9      173.2      66.3       50.2        2926   
146          0       97.0      173.5      65.4       53.0        2290   

     enginesize  boreratio  stroke  compressionratio  ...  CarBrand_nissan  \
15          209       3.62   3.390              8.00  ...                0   
9           131       3.13   3.400              7.00  ...                0   
100         120       3.33   3.470              8.50  ...                1   
132         121       3.54   3.070              9.31  ...                0   
68          183       3.58   3.640             21.50  ...                0   
95           97       3.15   3.290              9.40  ...                1   
159         110       3.27   3.350             22.50  ...                0   
162          98       3.19   3.030              9.00  ...                0   
147         108       3.62   2.640              9.00  ...                0   
182          97       3.01   3.400             23.00  ...                0   
191         136       3.19   3.400              8.50  ...                0   
16

In [21]:
y_train

66     18344.0
111    15580.0
153     6918.0
96      7499.0
38      9095.0
        ...   
106    18399.0
14     24565.0
92      6849.0
179    15998.0
102    14399.0
Name: price, Length: 164, dtype: float64

In [22]:
y_test

15     30760.000
9      17859.167
100     9549.000
132    11850.000
68     28248.000
95      7799.000
159     7788.000
162     9258.000
147    10198.000
182     7775.000
191    13295.000
164     8238.000
65     18280.000
175     9988.000
73     40960.000
152     6488.000
18      5151.000
82     12629.000
86      8189.000
143     9960.000
60      8495.000
101    13499.000
98      8249.000
30      6479.000
25      6692.000
16     41315.000
168     9639.000
195    13415.000
97      7999.000
194    12940.000
67     25552.000
120     6229.000
154     7898.000
202    21485.000
79      7689.000
69     28176.000
145    11259.000
55     10945.000
45      8916.500
84     14489.000
146     7463.000
Name: price, dtype: float64

### Scaling Procedure

In [23]:
##initialising Statndard scaler 

scaler = StandardScaler()



X_train_scaled = scaler.fit_transform(X_train)
X_train_scaled



array([[-0.72738032,  1.03708546,  0.07244984, ..., -0.42409446,
        -0.2548236 , -0.22645541],
       [-0.72738032,  1.54123134,  1.04391556, ..., -0.42409446,
        -0.2548236 , -0.22645541],
       [-0.72738032, -0.50896191, -0.36761583, ...,  2.35796522,
        -0.2548236 , -0.22645541],
       ...,
       [ 0.07863571, -0.71062026, -0.73295337, ..., -0.42409446,
        -0.2548236 , -0.22645541],
       [ 1.69066776,  0.70098821,  0.77821554, ...,  2.35796522,
        -0.2548236 , -0.22645541],
       [-0.72738032,  0.28086664,  0.86954992, ..., -0.42409446,
        -0.2548236 , -0.22645541]])

In [24]:
X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[-0.72738032,  0.80181738,  1.23488746, ..., -0.42409446,
        -0.2548236 , -0.22645541],
       [-0.72738032,  0.12962287,  0.33814986, ..., -0.42409446,
        -0.2548236 , -0.22645541],
       [-0.72738032, -0.25688897, -0.06040018, ..., -0.42409446,
        -0.2548236 , -0.22645541],
       ...,
       [-0.72738032, -0.71062026, -1.51344721, ..., -0.42409446,
        -0.2548236 , -0.22645541],
       [ 1.69066776, -0.47535218, -0.07700643, ..., -0.42409446,
        -0.2548236 , -0.22645541],
       [-0.72738032, -0.29049869, -0.05209705, ..., -0.42409446,
        -0.2548236 , -0.22645541]])

### Model Training & Evaluation


In [25]:
## Building a model to  Implement  regression algorithms:
def build_model(model,model_name,X_train, y_train,X_test, y_test):
    print("Running Model Name :",model)
    model.fit(X_train,y_train)
    results=model.predict(X_test)
    print("-------------------------------------------------------------------------------")



    return{
        "Model Name" : model_name+ "✅",
    "MSE" : mean_squared_error(y_test, results),
    "MAE" : mean_absolute_error(y_test, results),
    "R2 Score" : r2_score(y_test, results) }

In [26]:
## For Storing the Data 

results_list=[]

In [27]:
#1. LinearRegression

## Initialising the  Model 
lr_model=LinearRegression()

In [28]:
## Building and Evaluation of the model
lr_result = build_model(lr_model,"LinearRegression", X_train_scaled, y_train,X_test_scaled, y_test)
results_list.append(lr_result)
print(lr_result)

Running Model Name : LinearRegression()
-------------------------------------------------------------------------------
{'Model Name': 'LinearRegression✅', 'MSE': 7128546.783600219, 'MAE': 1763.5659861998556, 'R2 Score': 0.9097012452093493}


In [29]:
#2.DecisionTreeRegressor

## Initialising the  Model 
dtree_model= DecisionTreeRegressor()

In [30]:
## Building and Evaluation of the model
dtree_result = build_model(dtree_model,"DecisionTreeRegressor", X_train_scaled, y_train,X_test_scaled, y_test)
results_list.append(dtree_result)
print(dtree_result)

Running Model Name : DecisionTreeRegressor()
-------------------------------------------------------------------------------
{'Model Name': 'DecisionTreeRegressor✅', 'MSE': 8165288.502485098, 'MAE': 1857.4430975609757, 'R2 Score': 0.8965686265850042}


In [31]:
#3. RandomForestRegressor

## Initialising the  Model 
rfr_model= RandomForestRegressor()

In [32]:
## Building and Evaluation of the model
rfr_result = build_model(rfr_model,"RandomForestRegressor", X_train_scaled, y_train,X_test_scaled, y_test)
results_list.append(rfr_result)
print(rfr_result)

Running Model Name : RandomForestRegressor()
-------------------------------------------------------------------------------
{'Model Name': 'RandomForestRegressor✅', 'MSE': 3336327.21707568, 'MAE': 1256.122146341463, 'R2 Score': 0.9577380632516606}


In [33]:
#4. GradientBoostingRegressor

## Initialising the  Model 
gbr_model= GradientBoostingRegressor()

In [34]:
## Building and Evaluation of the model
gbr_result = build_model(gbr_model,"GradientBoostingRegressor", X_train_scaled, y_train,X_test_scaled, y_test)
results_list.append(gbr_result)
print(gbr_result)

Running Model Name : GradientBoostingRegressor()
-------------------------------------------------------------------------------
{'Model Name': 'GradientBoostingRegressor✅', 'MSE': 6080108.25561758, 'MAE': 1736.948940719558, 'R2 Score': 0.9229820296981566}


In [35]:
#5.Support Vector Regressor 

## Initialising the  Model 
svr_model= SVR()

In [36]:
## Building and Evaluation of the model
svr_result = build_model(svr_model,"Support Vector Regressor",X_train_scaled, y_train,X_test_scaled, y_test)
results_list.append(svr_result)
print(svr_result)

Running Model Name : SVR()
-------------------------------------------------------------------------------
{'Model Name': 'Support Vector Regressor✅', 'MSE': 86889433.59033778, 'MAE': 5701.535891309875, 'R2 Score': -0.10064616195306408}


### Comparing All results into an Dataframe 

In [37]:
results_df=pd.DataFrame(results_list)
results_df = results_df.sort_values(by='R2 Score', ascending=False).reset_index(drop=True)
results_df


Model Name           MSE          MAE  R2 Score
0      RandomForestRegressor✅  3.336327e+06  1256.122146  0.957738
1  GradientBoostingRegressor✅  6.080108e+06  1736.948941  0.922982
2           LinearRegression✅  7.128547e+06  1763.565986  0.909701
3      DecisionTreeRegressor✅  8.165289e+06  1857.443098  0.896569
4   Support Vector Regressor✅  8.688943e+07  5701.535891 -0.100646

### ✅ Best Working Model → RandomForestRegressor

#### Reasoning & Justification:

It has the highest R² score (0.9576) → explains ~95.8% of variance in the target variable.

Lowest errors: both MSE (3.34e+06) and MAE (1243.23) are the smallest among all models.

Indicates excellent generalization, capturing both linear and non-linear relationships effectively.

🔎 Why so good?
Random Forest is an ensemble method that averages predictions from many decision trees. This reduces overfitting (common in single Decision Trees) while maintaining flexibility for complex patterns.

### ❌ Worst Working Model → Support Vector Regressor (SVR)

#### Reasoning & Justification:

Negative R² score (-0.1006) → means the model is performing worse than a simple mean predictor.

Highest error values: MSE (8.68e+07) and MAE (5701.54), much larger than other models.

Indicates that SVR is not suitable for this dataset (at least with current parameters).

### 🔎 Why so bad?

SVR requires careful hyperparameter tuning (C, epsilon, kernel choice, scaling).

On high-dimensional or non-linearly separable data, the default SVR often underperforms.

It’s also sensitive to outliers, which may be affecting it here.

### Defining Feature Importance 

In [38]:
# Extract feature importances
f_imp= rfr_model.feature_importances_
feature_names = X_test.columns
feature_df = pd.DataFrame({'Feature': feature_names,'Importance': f_imp}).sort_values(by='Importance', ascending=False).reset_index(drop=True)



feature_df



Feature    Importance
0              enginesize  5.968549e-01
1              curbweight  2.399696e-01
2              highwaympg  4.800083e-02
3                carwidth  2.814988e-02
4              horsepower  2.665305e-02
..                    ...           ...
59  cylindernumber_twelve  3.969935e-06
60       enginetype_rotor  3.435595e-06
61     cylindernumber_two  2.333864e-06
62       enginetype_dohcv  6.346374e-07
63   cylindernumber_three  0.000000e+00

[64 rows x 2 columns]

plt.figure(figsize=(10,8))

sns.barplot(
    x="Feature",
    y="Importance",
    data=feature_df,
    )


plt.title('Feature Importance - Random Forest')
plt.xlabel("Features")
plt.ylabel("Importance",)
plt.tight_layout()
plt.xticks(rotation=90) 

plt.show()

### Grid Search CV for Hyperparameter Tuning 

In [43]:
# Initialize model
model = RandomForestRegressor(random_state=42)

# Create a grid of hyperparameters
param_grid = {
    'n_estimators': [100, 200, 300],           # Number of trees
    'max_depth': [None, 10, 20],               # Max depth of each tree
    'min_samples_split': [2, 5, 10]            # Minimum samples required to split a node
}

#  Setup Grid Search
grid_search = GridSearchCV(
    estimator=model,
    param_grid=param_grid,
    cv=5,                          # 5-fold cross-validation
    scoring='r2',                  # Evaluation metric
    n_jobs=-1                      # Use all available cores
)

# Fit the Grid Search
grid_search.fit(X_train_scaled, y_train)

# View best parameters and score
print("Best Parameters:", grid_search.best_params_)
print("Best R2 Score:", grid_search.best_score_)

# Predict using the best model
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test_scaled)

# Evaluate performance
print("Test R2 Score:", r2_score(y_test, y_pred))



Best Parameters: {'max_depth': 10, 'min_samples_split': 2, 'n_estimators': 300}
Best R2 Score: 0.8882275885740816
Test R2 Score: 0.9592352507777084


### Evaluating Model Performances

#### Base Model

In [61]:
##initialising 

base_model=RandomForestRegressor(random_state=42)


##fitting
base_model.fit(X_train,y_train)

#predicting & Evaluating 
y_pred=base_model.predict(X_test)
r2_base=r2_score(y_test,y_pred)

print("R2_Score:",r2_base)

R2_Score: 0.9584165239875437


#### Tuned Model

In [62]:
##initialising 

tuned_model=grid_search.best_estimator_

#predicting & Evaluating 
y_pred_tuned=tuned_model.predict(X_test)
r2_tuned=r2_score(y_test,y_pred_tuned)

print("R2_Score:",r2_tuned)

R2_Score: 0.9592582108394468


#### Base Model    VS     Tuned Model

In [72]:
## Before Tuning
print(f"Performance of Base Model(r2_score) : {r2_base:.4f}")

## After Tuning 
print(f"Performance of Tuned Model(r2_score) : {r2_tuned:.4f}")



## Differences 
print(f"Difference in base_score and Tuned_score : {diff:.4f}")

Performance of Base Model(r2_score) : 0.9584
Performance of Tuned Model(r2_score) : 0.9593
Difference in base_score and Tuned_score : 0.0008


### 🚗 Car Price Prediction – Management Insights

 The car price prediction model was built using the available independent variables, and it provides **valuable insights** into how prices vary with different features. This will help management in **design decisions, pricing strategy, and new market planning**.

### 🔑 Key Findings

* 📈 **Engine size, curb weight, horsepower, and car width** are the strongest drivers of higher car prices.
* ⛽ **Fuel efficiency (city mpg)** has an inverse relationship with price — higher efficiency usually means lower price.
* 🏆 **Random Forest Regressor** performed the best, achieving an **R² ≈ 0.96**, explaining \~96% of price variation.

### 💼 Business Implications

1. 🛠 **Car Design:** Adjusting technical specs like engine size & horsepower can position cars in premium or budget segments.
2. 🎯 **Market Strategy:** Balance between performance features and fuel efficiency to attract target customers.
3. 🌍 **New Market Entry:** Use the model to simulate pricing dynamics in unfamiliar markets and set competitive prices.

---

## ✅ **Conclusion**

This ML regression model is not only a **highly accurate predictor of car prices** but also a **strategic decision-making tool**. It enables management to:
✔️ Understand the impact of design features on price
✔️ Optimize business strategies for different market segments
✔️ Confidently plan pricing in new markets

🚀 **In short:** The model bridges **data-driven insights** with **business strategy**, giving the company a strong competitive edge.